File: cpp/iprocess.cpp
// g++ -o iprocess iprocess.cpp -I/usr/local/include -I/usr/local/include/opencv -L/usr/local/lib/ -lopencv_core -lopencv_highgui -lopencv_imgproc -lopencv_imgcodecs -std=c++11

#include <stdio.h>
#include <string.h>
#include <time.h>
#include <iostream>
#include <unistd.h>
#include <string>
#include <opencv2/opencv.hpp>
#include <sys/time.h>


using namespace cv;
using namespace std;


void process(Mat &img)
{
	// change s
	img.convertTo(img, CV_32FC3, 1.0/255);
	cvtColor(img, img, COLOR_BGR2HLS);
	
	int rows = img.rows;
	int cols = img.cols;
	int channels = img.channels();
	int step = img.step;
	step /= sizeof(float);
	float* data = (float*)img.data;
	for(int i=0;i<rows;i++)
		for(int j=0;j<cols;j++)
		{
			int pos = i*step + j*channels + 2;
			float res = data[pos] * 1.5;
			data[pos] = res > 1 ? 1 : res;
		}

	// threshold(img, img, 1, 1, THRESH_TRUNC);
	cvtColor(img, img, COLOR_HLS2BGR);
	img.convertTo(img, CV_8UC3, 255);

	// blur
	medianBlur(img, img, 5);
	GaussianBlur(img, img, Size(3,3), 1);
}


void read_and_write(string image_name)
{
	Mat img = imread(image_name);
	process(img);
	if (img.empty()) {
		cout << "Failed in processing image." << endl;
		return;
	}
	imwrite("test-c.bmp", img);
}


int main()
{
	string image_name = "./test.bmp";
	read_and_write(image_name);

	return 0;
}

File: cpp/iprocess.py
import os
import cv2
import numpy as np
from datetime import datetime


def hls_trans(image):
    hlsImg = image.astype(np.float32)
    hlsImg = hlsImg / 255.0
    hlsImg = cv2.cvtColor(hlsImg, cv2.COLOR_BGR2HLS)
    hlsImg[:, :, 2] = 1.5 * hlsImg[:, :, 2]
    hlsImg[:, :, 2][hlsImg[:, :, 2] > 1] = 1
    hlsImg = cv2.cvtColor(hlsImg, cv2.COLOR_HLS2BGR)
    hlsImg = hlsImg * 255
    image = hlsImg.astype(np.uint8)
    return image

# wrong approach
def hls_trans_RGB(image):
    hlsImg = image.astype(np.float32)
    hlsImg = hlsImg / 255.0
    hlsImg = cv2.cvtColor(hlsImg, cv2.COLOR_RGB2HLS)
    hlsImg[:, :, 2] = 1.5 * hlsImg[:, :, 2]
    hlsImg[:, :, 2][hlsImg[:, :, 2] > 1] = 1
    hlsImg = cv2.cvtColor(hlsImg, cv2.COLOR_HLS2RGB)
    hlsImg = hlsImg * 255
    image = hlsImg.astype(np.uint8)
    return image


def hls_trans_(image):
    hlsImg = image.astype(np.float32)
    hlsImg = hlsImg / 255.0
    hlsImg = cv2.cvtColor(hlsImg, cv2.COLOR_BGR2HLS)
    hlsImg[:, :, 2] = 1.5 * hlsImg[:, :, 2]
    # hlsImg[:, :, 2][hlsImg[:, :, 2] > 1] = 1
    _, hlsImg[:, :, 2] = cv2.threshold(hlsImg[:, :, 2], 1, 1, cv2.THRESH_TRUNC)
    hlsImg = cv2.cvtColor(hlsImg, cv2.COLOR_HLS2BGR)
    hlsImg = hlsImg * 255
    image = hlsImg.astype(np.uint8)
    return image


def process(image):
    image = hls_trans(image)
    image = cv2.medianBlur(image, 5)
    image = cv2.GaussianBlur(image, (3,3), 1)
    return image


if __name__ == "__main__":
    a = 'test.bmp'
    img = cv2.imread(a)

    # time1 = datetime.now()
    # for i in range(1000):
    #     i1 = hls_trans(img)
    # time2 = datetime.now()
    # print(time2 - time1)

    # time1 = datetime.now()
    # for i in range(1000):
    #     i2 = hls_trans_(img)
    # time2 = datetime.now()
    # print(time2 - time1)


    # i1 = process(img)
    # cv2.imwrite('test-py1.bmp', i1)

    # i2 = hls_trans_RGB(img)
    # cv2.imwrite('test-py1-rgb.bmp', i2)


    a = 'test-c.bmp'
    b = 'test-py1.bmp'
    i1 = cv2.imread(a)
    i2 = cv2.imread(b)


    count = 0
    h, w, c = i1.shape
    for i in range(h):
        for j in range(w):
            for k in range(c):
                if abs(i1[i][j][k] - i2[i][j][k]) > 0:
                    count += 1
                    print(i1[i][j][k], i2[i][j][k])
    print(count)